### DISCUSSION TITLE
(dis)agreement between solvers

### DISCUSSION THREAD

**Discussion by mihai-sysbio** [LCSB-BioCore/COBREXA.jl#815] [Q&A] [answered]
I've been running a knockout screen of 600+ genes across 50 tissue models. Using COBREXA, I've leveraged its use of the JuMP framework to test over 20 solvers. The main take-away is that HiGHS (free, open-source) gave identical results to Gurobi and Mosek (commercial, academic licenses), and almost identical with Tulip (which was slightly worse and about 50% longer). 
There is one puzzling bit though: OSQP, a different solver, has managed to find flux vectors with 1.5x higher flux carried by the objective function than what the 3 other robust solvers have. Does anyone have any experience/thoughts on this?

**Comment by exaexa** [comment, marked as answer]
Hi,
yes this is common. I'm currently finding workforce for making a rigorous benchmark of all this stuff (actually I ran quite some already with FROG models, which is a nice curated dataset, but we need someone to actually interpret it...).

The current findings:
- HiGHS is pretty good, the solver that I generally liked most was SCIP
- Gurobi sometimes overshoots optima by tolerance so much that it then fails at 99%-optimum FVA (this is a pretty good test for tolerance handling btw)
- CPLEX is mostly OKish but not open
- OSQP has terrible problems with tolerances sometimes reporting success despite having unsatisfied constraints (I assume that might be your case)
- Tulip.jl is super cool for small problems, Clarabel.jl is surprisingly cool for quadratic stuff (sometimes you need to give it a bit of a poke by initializing it by another linear solver's output, but then it optimizes QPs very nicely)
- GLPK is super conservative w.r.t. tolerances, thus sometimes not finding a solution.

Linear solving is a wild business.

I'm attaching a picture with performances (completely uncompensated/uncurated) from the mentioned benchmark. HiGHS ain't bad. I kicked out some solvers (including OSQP) for producing errors. Also, I unfortunately didn't know SCIP is a thing by then.

Hope this helps. :D

[fba-all.pdf](https://github.com/LCSB-BioCore/COBREXA.jl/files/13860424/fba-all.pdf)

BTW this list is quite useful because it's usually the most updated :D https://jump.dev/JuMP.jl/stable/installation/#Supported-solvers

**Reply by mihai-sysbio** [reply]
This is extremely reassuring! Thanks for sharing those benchmark results. The JuMP solver list was my starting point, but I didn't get Clp to work. One thing that becomes obvious is the use-case for the solver: solving a big problem (Human-GEM) tens of times vs a small problem (iDS372) thousands of times. It might be useful to have a look over Daniel Machado's preprint [A benchmark of optimization solvers for genome-scale metabolic modeling](https://www.biorxiv.org/content/10.1101/2023.04.11.536343v2) and the associated repository https://github.com/cdanielmachado/solver_benchmark

**Reply by exaexa** [reply]
uuh that's a good repo :D

Do you think we could just throw in more data from our JuMP attempts here?

**Reply by mihai-sysbio** [reply]
I have no idea, it would be better to reach out to @cdanielmachado directly

**Reply by cdanielmachado** [reply]
Hi, just let me know how I can help :)

**Reply by exaexa** [reply]
Hey all, just an update here. We've kinda pre-finished the FROG effort, with curating and seeing quite a lot of various numerical errors among solvers. Mostly these are the "common type" of error amplification problems in the models -- the cause with the absolutely biggest prevalence is simply using small and big numbers too close together.

I didn't formalize any general advice at this point, but the current advice&status is:
- check the significant magnitude of all values that go to the solver, and make very sure all of them are at most around 10^4 apart
- try with one IPM solver and one simplex-y solver; if there's some numerical issue it's gonna be super visible on the difference between the methods
- there is no clear winner among solvers

Possibly also of interest: https://github.com/ds4dm/Tulip.jl/issues/94

**Reply by exaexa** [reply]
@cdanielmachado btw I see your [paper](https://doi.org/10.1128/msystems.00833-23) came out literally 2 weeks after we were discussing this, looks very nice. Did you quantify the solver (dis)agreement on particular solutions there somehow? Asking mainly to know if there's a good metric for that-- the only one I managed to think of was literally going over the constraints manually, and making statistics on how much error got accumulated in which one.

**Reply by cdanielmachado** [reply]
By disagreement do you mean different values for the objective function or different flux distributions ?

I did not observe differences in the final value of the objective function for the different solvers. Sometimes the solvers return different flux distributions for the same objective value, but that's just due to degeneracy of the optimal solution.

**Reply by exaexa** [reply]
> By disagreement do you mean different values for the objective function or different flux distributions ?

Objective values and some other things such as feasible ranges.

For example in FROG curation, the differences in objective value existed, but most were tolerable (as in 1e-6). Instead we saw pretty wild differences in e.g. ranges from FVA (10% relative tolerance was sometimes insufficient) and sometimes even in single gene knockout FBAs. I don't have any actual output at hand that would quantify these problems, since FROG was mainly about making sure that the models solve right, or producing an exclamation mark in cases where they don't.

I'll wait until the publication gets out (hopefully soon) and start poking at this more precisely.

(Fun story with Gurobi: in a few cases I saw that it somehow "rounded up" the optimal objective value (no actual idea what happened there -- for example, it was supposed to return objective value 1.0, but it returned 1.00005), and then it refused to find any solution for a FVA ran at 100% objective value, insisting there's no feasible solution with objective >=1.00005 (and thus contradicting its first run).)


PS. edited to actually answer the question clearly, sorry for mess :D

**Reply by exaexa** [reply]
PS FROG links :frog: 

- https://www.ebi.ac.uk/biomodels/curation/fbc
- some software here: https://github.com/LCSB-BioCore/FBCModelTests.jl/tree/master#frog
- frog spec with some rationale on what's done there https://github.com/LCSB-BioCore/frog-specification

**Reply by cdanielmachado** [reply]
Interesting, I had never heard about FROG before. Is it supposed to replace or complement MEMOTE ? 

Are the results in the report evaluating the quality of the model or the solver ?

**Reply by exaexa** [reply]
> replace MEMOTE

no, it's complementary. Memote is more about having annotations and making sure the model contains no semantic surprise ("magic ATP from nowhere!") etc. FROG is mostly about the solver interaction and software interpretation (if I run this on 2 completely different software stacks, am I gonna get the same result?). Unfortunately the FROG task cannot be solved by static tests (unless we build models with known ground truth from the beginning), so it's mainly about checking whether different implementations agree with what authors intended to show.

**Reply by exaexa** [reply]
re "model or the solver" -- depends, we got models which solve reliably with solver A but not B, and others which go better with B. I guess that avoidance of numeric issues is kinda model responsibility, but can you reliably avoid all numeric issues in all future solver versions? (no)

**Reply by cdanielmachado** [reply]
I see, that's an interesting validation to have. Although i would say the job of the model is to represent the biology. Dealing with numerical issues should be the responsibility of the simulation methods. I'd say we don't really "solve a model", we solve a particular formulation of a particular simulation method. :)

**Reply by exaexa** [reply]
Yeah, "validation" would say it best.

The point with the model interpretation is valid but the scale is huge, in extremes you can go for "interpret as pure data" (and just publish the genome untainted by COBRA view, letting everyone build the model themselves) or "interpret as program" (which gets you perfect reproducibility for the predictions etc, but good luck guessing what all the magic constants and reactions do there). We hover somewhere in the middle in the high utility area... :)

**Reply by mihai-sysbio** [reply]
Interesting how this turned into an exploration of what a _model_ is. _Representation of biology_ indeed, but does it work only for this one dataset? or this one research question? or only when I run it on _my_ machine? In my view the general aim is for the model to perform "reliably"/"consistently": for more users (including in different sofware environments) and for more research questions. Every bit of improvement counts 💪🏻

**Reply by exaexa** [reply]
@mihai-sysbio the "reliably/consistently" is basically in favor of "model as tested software"; you can't really do any reliability or consistency test on "base data".

OTOH since software is basically a function (or many functions) that transforms tons of parameters into some kind of result, this view for me kinda puts a limit on the re-usability of the model data (what if you want to do something that the authors didn't ever think of).

So maybe (to avoid the "reliable blackbox" view of models), we could go with a software where all code has some attached biological annotation? e.g. "we multiply by 1.23456 here because that's what was measured by Guy et al [123]"

...or maybe data with precisely standardized ways of interpretation? (Think of SBML, but where they actually standardized the interpretation. :D )
